### PR TITLE
[IMPROVE] Replace livechat inquiry dialog with preview room

### LINKED ISSUE #13990: Replace "Take Conversation" model with the new Join Room UX
Allow agents to preview the channel before taking it.

image

We should use the same new UX of the joining room, but replace the button label with "Take it".

### PR DESCRIPTION
**Pull request by renatobecker**
Closes #13990.

This PR replaces the current dialog used to take livechat inquiries with the new `Join Room` approach, as shown below:

Screen Shot 2019-04-02 at 20 33 48

In addition: When the livechat conversation is closed, only the client that requested the action is notifyed, if any other client is accessing the room, nothing happens.

### CODE REVIEW

File: app/utils/client/lib/roomTypes.js [hunk 1]
```diff
@@ -105,16 +112,16 @@ export const roomTypes = new class RocketChatRoomTypes extends RoomTypesCommon {
 		}
 		return this.roomTypes[roomType].showJoinLink(roomId);
 	}
-	getNotSubscribedTpl(roomId) {
+	getTemplate(roomId, templateName) {
```
Comment (sampaiodiego) [inline app/utils/client/lib/roomTypes.js]: I didn't like changing this function to make it more generic.. It was made specific on purpose.. I would rather create a new one to get the `readOnlyTpl`..

to avoid duplicating code that finds the room I'd then create a private method/function to do that.

File: app/livechat/client/views/app/livechatReadOnly.html [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<template name="livechatReadOnly">
+    {{#if roomOpen}}
```
[the remark this answers is not in the document]
Comment (renatobecker) [reply app/livechat/client/views/app/livechatReadOnly.html]: Done.

File: app/livechat/client/views/app/livechatReadOnly.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+});
+
+Template.livechatReadOnly.events({
+	'click .js-take-it'(event, instance) {
+		event.preventDefault();
+		event.stopPropagation();
+
+		const inquiry = instance.inquiry.get();
+		const { _id } = inquiry;
+		call('livechat:takeInquiry', _id, (error /* ,result */) => {
```
[the remark this answers is not in the document]
Comment (renatobecker) [reply app/livechat/client/views/app/livechatReadOnly.js]: Done.

File: app/livechat/lib/LivechatRoomType.js [hunk 4]
```diff
@@ -85,4 +84,19 @@ export default class LivechatRoomType extends RoomTypeConfig {
 				return '';
 		}
 	}
+
+	readOnly(roomId, user) {
+		const room = ChatRoom.findOne({ _id: roomId }, { fields: { open: 1, servedBy: 1 } });
+		if (!(room && room.open === true)) {
```
[the remark this answers is not in the document]
Comment (renatobecker) [reply app/livechat/lib/LivechatRoomType.js]: Done.

File: app/livechat/lib/LivechatRoomType.js [hunk 5]
```diff
@@ -85,4 +84,19 @@ export default class LivechatRoomType extends RoomTypeConfig {
...
+			return true;
+		}
+
+		const inquiry = LivechatInquiry.findOne({ rid: roomId }, { fields: { status: 1 } });
+		if (inquiry && inquiry.status === 'open') {
+			return true;
+		}
+
+		return !((room.servedBy && room.servedBy._id === user._id) || hasPermission('view-livechat-rooms'));
```
[the remark this answers is not in the document]
Comment (renatobecker) [reply app/livechat/lib/LivechatRoomType.js]: Done.

File: app/ui-message/client/messageBoxReadOnly.html [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<template name="messageBoxReadOnly" args="_id">
+	{{#if customTemplate}}
```
[the remark this answers is not in the document]
Comment (renatobecker) [reply app/ui-message/client/messageBoxReadOnly.html]: Done.

File: app/livechat/server/startup.js [hunk 7]
```diff
@@ -20,6 +21,19 @@ Meteor.startup(() => {
 		return room && room.t === 'l' && extraData && extraData.visitorToken && room.v && room.v.token === extraData.visitorToken;
 	});
 
+	addRoomAccessValidator(function(room, user) {
+		if (settings.get('Livechat_Routing_Method') !== 'Guest_Pool') {
+			return;
+		}
+
+		if (!(room && room.t === 'l' && user)) {
```
[the remark this answers is not in the document]
Comment (renatobecker) [reply app/livechat/server/startup.js]: Done.

File: app/utils/client/lib/roomTypes.js (hunk 1 above)
Comment (renatobecker) [reply app/utils/client/lib/roomTypes.js]: Done.

**Comment by renatobecker** [comment]
@tassoevan, when the agent clicks on the `Take it` button, the composer appears, but they can't send a message because of the `this.input` is not assigned.

File: app/utils/client/lib/roomTypes.js [hunk 8]
```diff
@@ -38,23 +38,30 @@ export const roomTypes = new class RocketChatRoomTypes extends RoomTypesCommon {
 			rid: roomId,
 		}).count() > 0;
 	}
-	readOnly(rid, user) {
+	readOnly(roomId, user) {
```
Comment (ggazzo) [inline app/utils/client/lib/roomTypes.js]: why did you change the variable name? we are trying to use the same name `rid` for all points

Comment (ggazzo) [inline app/utils/client/lib/roomTypes.js]:
```suggestion
	readOnly(rid, user) {
```

File: app/utils/client/lib/roomTypes.js [hunk 9]
```diff
@@ -38,23 +38,30 @@ export const roomTypes = new class RocketChatRoomTypes extends RoomTypesCommon {
...
 		const fields = {
 			ro: 1,
+			t: 1,
 		};
 		if (user) {
 			fields.muted = 1;
 		}
 		const room = ChatRoom.findOne({
-			_id: rid,
+			_id: roomId,
```
Comment (ggazzo) [inline app/utils/client/lib/roomTypes.js]:
```suggestion
			_id: rid,
```

File: app/utils/client/lib/roomTypes.js [hunk 10]
```diff
@@ -38,23 +38,30 @@ export const roomTypes = new class RocketChatRoomTypes extends RoomTypesCommon {
...
+		if (room && roomType && this.roomTypes[roomType] && this.roomTypes[roomType].readOnly) {
+			return this.roomTypes[roomType].readOnly(roomId, user);
+		}
+
 		if (!user) {
 			return room && room.ro;
 		}
 		const userOwner = RoomRoles.findOne({
-			rid,
+			rid: roomId,
```
Comment (ggazzo) [inline app/utils/client/lib/roomTypes.js]:
```suggestion
			rid,
```

File: app/utils/client/lib/roomTypes.js [hunk 11]
```diff
@@ -117,6 +124,18 @@ export const roomTypes = new class RocketChatRoomTypes extends RoomTypesCommon {
 		return this.roomTypes[roomType].notSubscribedTpl;
 	}
 
+	getReadOnlyTpl(roomId) {
```
Comment (ggazzo) [inline app/utils/client/lib/roomTypes.js]:
```suggestion
	getReadOnlyTpl(rid) {
```

File: app/utils/client/lib/roomTypes.js [hunk 12]
```diff
@@ -117,6 +124,18 @@ export const roomTypes = new class RocketChatRoomTypes extends RoomTypesCommon {
...
+		const room = ChatRoom.findOne({ _id: roomId }, { fields: { t: 1 } });
```
Comment (ggazzo) [inline app/utils/client/lib/roomTypes.js]:
```suggestion
		const room = ChatRoom.findOne({ _id: rid }, { fields: { t: 1 } });
```

Comment (ggazzo) [inline app/utils/client/lib/roomTypes.js]: why not pass `t: { $exists: 1 }` if you agree plz check because there are more places like that

File: app/utils/client/lib/roomTypes.js [hunk 13]
```diff
@@ -38,23 +38,30 @@ export const roomTypes = new class RocketChatRoomTypes extends RoomTypesCommon {
...
 		}
 		const room = ChatRoom.findOne({
-			_id: rid,
+			_id: roomId,
 		}, {
 			fields,
 		});
+
+		const roomType = room.t;
+		if (room && roomType && this.roomTypes[roomType] && this.roomTypes[roomType].readOnly) {
```
Comment (ggazzo) [inline app/utils/client/lib/roomTypes.js]: I dont think you need to check this `room` as you already made `const roomType = room.t;` or change the tests place

File: app/livechat/lib/LivechatRoomType.js [hunk 14]
```diff
@@ -87,6 +86,20 @@ export default class LivechatRoomType extends RoomTypeConfig {
 		}
 	}
 
+	readOnly(roomId, user) {
```
Comment (ggazzo) [inline app/livechat/lib/LivechatRoomType.js]: rid <3

File: app/livechat/client/views/app/livechatReadOnly.js [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+});
+
+Template.livechatReadOnly.events({
+	async 'click .js-take-it'(event, instance) {
+		event.preventDefault();
+		event.stopPropagation();
+
+		const inquiry = instance.inquiry.get();
+		const { _id } = inquiry;
+		try {
```
Comment (ggazzo) [inline app/livechat/client/views/app/livechatReadOnly.js]: if I'm not mistaken `call` already has a `handleError`you dont need to handle by yourself

File: app/livechat/server/publications/livechatInquiries.js [hunk 16]
```diff
@@ -16,5 +16,27 @@ Meteor.publish('livechat:inquiry', function() {
 		status: 'open',
 	};
 
-	return LivechatInquiry.find(query);
+	if (_id) {
+		Object.assign(query, { _id });
```
Comment (ggazzo) [inline app/livechat/server/publications/livechatInquiries.js]: why not 
```javascript
const query = {
		agents: this.userId,
		status: 'open',
		...(_id && { _id }),
}	
//or just 
query._id = _id;
```
